src/mesh/mesh_dual_weighted_residual_error.cpp

```diff
@@ -0,0 +1,388 @@
+#include <vector>
+#include <iostream>
+#include <fstream>
+
+#include <Epetra_RowMatrixTransposer.h>
+
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/grid/tria.h>
+#include <deal.II/distributed/shared_tria.h>
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/distributed/solution_transfer.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/data_out.h>
+
+#include "parameters/all_parameters.h"
+
+#include "dg/dg.h"
+#include "mesh_error_estimate.h"
+#include "functional/functional.h"
+#include "physics/physics.h"
+#include "linear_solver/linear_solver.h"
+#include "post_processor/physics_post_processor.h"
+
+namespace PHiLiP {
+
+template <int dim, typename real, typename MeshType>
+MeshErrorEstimateBase<dim, real, MeshType> :: ~MeshErrorEstimateBase(){}
+
+// constructor
+template <int dim, int nstate, typename real, typename MeshType>
+DualWeightedResidualError<dim, nstate, real, MeshType>::DualWeightedResidualError(std::shared_ptr< DGBase<dim, real, MeshType> > dg):
+    solution_coarse(dg->solution),
+    adjoint_state(AdjointStateEnum::coarse),
+    mpi_communicator(MPI_COMM_WORLD),
+    pcout(std::cout, dealii::Utilities::MPI::this_mpi_process(mpi_communicator)==0)
+{
+    // storing the original FE degree distribution
+    coarse_fe_index.reinit(dg->triangulation->n_active_cells());
+
+    // Create functional
+    functional = FunctionalFactory<dim,nstate,real,MeshType>::create_Functional(dg->all_parameters->functional_param, dg);
+
+    // looping over the cells
+    for(auto cell = dg->dof_handler.begin_active(); cell != dg->dof_handler.end(); ++cell)
+        if(cell->is_locally_owned())
+            coarse_fe_index[cell->active_cell_index()] = cell->active_fe_index();
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+real DualWeightedResidualError<dim, nstate, real, MeshType>::total_dual_weighted_residual_error(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
```

Comment:
Similar to MeshAdaptation, every function in this class takes dg as an argument. Can the class not just contain a pointer to dg as a member variable?

Reply:
Same as above. It's a member now.